src/charm.py

```diff
@@ -145,155 +193,63 @@ def _patch_service(self, name: str, ro_port: int, rw_port: int) -> None:
-            if version := MySQLRouter.get_version(container):
-                self.unit.set_workload_version(version)
-
-    def _on_leader_elected(self, _) -> None:
-        """Handle the leader elected event.
-
-        Patch existing k8s service to include read-write and read-only services.
-        """
-        # Create the read-write and read-only services
-        try:
-            self._patch_service(f"{self.app.name}", ro_port=6447, rw_port=6446)
-        except ApiError:
-            logger.exception("Failed to patch k8s service")
-            self.unit.status = BlockedStatus("Failed to patch k8s service")
-            return
-
-    def _on_mysql_router_pebble_ready(self, _) -> None:
-        """Handle the mysql-router pebble ready event."""
-        if self._bootstrap_mysqlrouter():
-            self.unit.status = ActiveStatus()
-
-    def _on_peer_relation_changed(self, _) -> None:
-        """Handle the peer relation changed event.
-
-        Bootstraps mysqlrouter if the relations exist, but pebble_ready event
-        fired before the requires relation was formed.
-        """
+    def _reconcile_database_relations(self, event=None) -> None:
+        """Handle database requires/provides events."""
+        logger.debug(
+            "State of reconcile "
+            f"{self.unit.is_leader()=}, "
+            f"{isinstance(self.workload, workload.AuthenticatedWorkload)=}, "
+            f"{self.database_requires.relation and self.database_requires.relation.is_breaking(event)=}, "
+            f"{self.workload.container_ready=}"
+        )
         if (
-            isinstance(self.unit.status, WaitingStatus)
-            and self.app_peer_data.get(MYSQL_DATABASE_CREATED)
-            and self._bootstrap_mysqlrouter()
+            self.unit.is_leader()
+            and isinstance(self.workload, workload.AuthenticatedWorkload)
+            and self.workload.container_ready
         ):
-            self.unit.status = ActiveStatus()
-
-        if self.unit.is_leader():
-            num_units_bootstrapped = sum(
-                1
-                for _ in self.peers.units.union({self.unit})
-                if self.unit_peer_data.get(UNIT_BOOTSTRAPPED)
+            self.database_provides.reconcile_users(
+                event=event,
+                event_is_database_requires_broken=self.database_requires.relation.is_breaking(
+                    event
+                ),
+                router_endpoint=self._endpoint,
+                shell=self.workload.shell,
             )
-            self.app_peer_data[NUM_UNITS_BOOTSTRAPPED] = str(num_units_bootstrapped)
+        if (
+            isinstance(self.workload, workload.AuthenticatedWorkload)
+            and self.workload.container_ready
+        ):
+            if self.database_requires.relation.is_breaking(event):
+                self.workload.disable()
+            else:
+                self.workload.enable(tls=self.tls.certificate_saved)
```

Comment:
do we need exception handling here in case `wait_until_mysql_router_ready` throws an assertion error?